File: src/domain/entities/economicKind.ts
/**
 * الأنواع الاقتصادية — جدول spec/02-accounting.md حرفيًا.
 *
 * القاعدة الحاكمة: **وارد الحساب لا يحدد وحده النوع الاقتصادي.**
 * اتجاه السيولة (وارد/صادر) حقيقة بنكية؛ النوع الاقتصادي قرار محاسبي.
 * التصنيف الآلي يقترح، وتأكيد المستخدم يحمي اختياره من الكتابة الآلية اللاحقة.
 */

export type EconomicKind =
  // وارد — يزيد الدخل
  | 'salary' // مرتب
  | 'bonus' // بونص
  | 'commission' // عمولة
  | 'overtime' // أوفر تايم
  | 'freelance' // فري لانس
  | 'personal_sale' // بيع شخصي
  // وارد — لا يزيد الدخل
  | 'loan_received' // قرض مستلم → دين عليك يزيد
  | 'debt_collected' // تحصيل دين لك → دين لك يقل
  | 'custody_received' // أمانة مستلمة → أمانة للآخر تزيد
  // صادر — مصروف
  | 'purchase' // شراء / فاتورة
  | 'support_gift' // دعم / هدية إلى مصر
  | 'fee' // رسوم (مصروف مستقل مرتبط بحركة)
  // صادر — ليس مصروفًا
  | 'loan_granted' // قرض ممنوح → دين لك يزيد
  | 'debt_repaid' // سداد دين عليك → دين عليك يقل
  | 'custody_returned' // رد أمانة → الأمانة للآخر تقل
  // داخلي
  | 'internal_transfer' // بين محافظي / سحب إلى محفظة كاش
  // استثمار
  | 'asset_buy' // شراء أصل استثماري
  | 'asset_sell' // بيع أصل استثماري
  // غير محدد بعد
  | 'unclassified'

/** اتجاه السيولة كما هو في الكشف — حقيقة بنكية لا تُستنتج. */
export type LiquidityDirection = 'in' | 'out' | 'internal'

/** أثر النوع على حساب الشخص. */
export type PersonEffect =
  | 'none'
  | 'receivable_up' // لك عنده يزيد
  | 'receivable_down' // لك عنده يقل
  | 'payable_loan_up' // له عندك (قرض) يزيد
  | 'payable_loan_down'
  | 'payable_custody_up' // له عندك (أمانة) يزيد
  | 'payable_custody_down'
  | 'beneficiary_info' // معلومة مستفيد فقط، بلا التزام

export interface EconomicKindRule {
  readonly kind: EconomicKind
  readonly label: string
  readonly liquidity: LiquidityDirection
  /** هل يزيد الدخل الاقتصادي للفترة؟ */
  readonly countsAsIncome: boolean
  /** هل يزيد المصروف الشخصي؟ (بنصيب المستخدم فقط بعد فصل المستحق على الآخرين) */
  readonly countsAsPersonalExpense: boolean
  readonly personEffect: PersonEffect
}

const RULES: Record<EconomicKind, EconomicKindRule> = {
  salary: { kind: 'salary', label: 'مرتب', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },
  bonus: { kind: 'bonus', label: 'بونص', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },
  commission: { kind: 'commission', label: 'عمولة', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },
  overtime: { kind: 'overtime', label: 'أوفر تايم', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },
  freelance: { kind: 'freelance', label: 'فري لانس', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },
  personal_sale: { kind: 'personal_sale', label: 'بيع شخصي', liquidity: 'in', countsAsIncome: true, countsAsPersonalExpense: false, personEffect: 'none' },

  loan_received: { kind: 'loan_received', label: 'قرض مستلم', liquidity: 'in', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'payable_loan_up' },
  debt_collected: { kind: 'debt_collected', label: 'تحصيل دين لك', liquidity: 'in', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'receivable_down' },
  custody_received: { kind: 'custody_received', label: 'أمانة مستلمة', liquidity: 'in', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'payable_custody_up' },

  purchase: { kind: 'purchase', label: 'شراء / فاتورة', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: true, personEffect: 'receivable_up' },
  support_gift: { kind: 'support_gift', label: 'دعم / هدية', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: true, personEffect: 'beneficiary_info' },
  fee: { kind: 'fee', label: 'رسوم', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: true, personEffect: 'none' },

  loan_granted: { kind: 'loan_granted', label: 'قرض ممنوح', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'receivable_up' },
  debt_repaid: { kind: 'debt_repaid', label: 'سداد دين عليك', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'payable_loan_down' },
  custody_returned: { kind: 'custody_returned', label: 'رد أمانة', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'payable_custody_down' },

  internal_transfer: { kind: 'internal_transfer', label: 'تحويل داخلي', liquidity: 'internal', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'none' },

  asset_buy: { kind: 'asset_buy', label: 'شراء أصل', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'none' },
  // بيع الأصل: لا دخل معيشة من كامل الحصيلة — الربح المحقق فقط، ويُحسب في الاستثمار
  asset_sell: { kind: 'asset_sell', label: 'بيع أصل', liquidity: 'in', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'none' },

  unclassified: { kind: 'unclassified', label: 'غير محدد', liquidity: 'out', countsAsIncome: false, countsAsPersonalExpense: false, personEffect: 'none' },
}

export function ruleFor(kind: EconomicKind): EconomicKindRule {
  const rule = RULES[kind]
  if (!rule) throw new Error(`نوع اقتصادي غير معروف: ${kind}`)
  return rule
}

export const ALL_ECONOMIC_KINDS = Object.keys(RULES) as EconomicKind[]

export const countsAsIncome = (kind: EconomicKind): boolean => ruleFor(kind).countsAsIncome
export const countsAsPersonalExpense = (kind: EconomicKind): boolean =>
  ruleFor(kind).countsAsPersonalExpense
export const liquidityOf = (kind: EconomicKind): LiquidityDirection => ruleFor(kind).liquidity
export const personEffectOf = (kind: EconomicKind): PersonEffect => ruleFor(kind).personEffect

/**
 * هل اتجاه السيولة المُلاحَظ في الكشف متسق مع النوع الاقتصادي المقترح؟
 * يُستخدم للتحذير عند المراجعة، لا لرفض تلقائي —
 * لأن الكشف يعرف الاتجاه فقط ولا يعرف النية.
 */
export function isConsistentWithObservedDirection(
  kind: EconomicKind,
  observed: 'in' | 'out',
): boolean {
  const expected = liquidityOf(kind)
  if (expected === 'internal') return true // التحويل الداخلي له طرفان
  return expected === observed
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/EconomicKind.kt
package app.masroufy.core

import app.masroufy.core.EconomicKind.*

/**
 * الأنواع الاقتصادية — نقل `src/domain/entities/economicKind.ts` (جدول spec/02 حرفيًا).
 * **وارد الحساب لا يحدد وحده النوع الاقتصادي**: اتجاه السيولة حقيقة بنكية، والنوع قرار محاسبي.
 * `wire` = الاسم المتخزن في فايربيز — ما يتغيرش.
 */
enum class EconomicKind(val wire: String) {
    // وارد — بيزود الدخل
    SALARY("salary"), BONUS("bonus"), COMMISSION("commission"), OVERTIME("overtime"), FREELANCE("freelance"),
    PERSONAL_SALE("personal_sale"),
    // وارد — ما بيزودش الدخل
    LOAN_RECEIVED("loan_received"), DEBT_COLLECTED("debt_collected"), CUSTODY_RECEIVED("custody_received"),
    // صادر — مصروف
    PURCHASE("purchase"), SUPPORT_GIFT("support_gift"), FEE("fee"),
    // صادر — مش مصروف
    LOAN_GRANTED("loan_granted"), DEBT_REPAID("debt_repaid"), CUSTODY_RETURNED("custody_returned"),
    // داخلي
    INTERNAL_TRANSFER("internal_transfer"),
    // استثمار
    ASSET_BUY("asset_buy"), ASSET_SELL("asset_sell"),
    // لسه ما اتحددش
    UNCLASSIFIED("unclassified");

    companion object {
        fun fromWire(wire: String): EconomicKind =
            entries.firstOrNull { it.wire == wire } ?: throw IllegalArgumentException(uiText(TextKey.KIND_UNKNOWN_WIRE, wire))
    }
}

enum class Direction(val wire: String) {
    IN("in"), OUT("out");

    companion object {
        fun fromWire(wire: String): Direction = entries.first { it.wire == wire }
    }
}

enum class Liquidity(val wire: String) { IN("in"), OUT("out"), INTERNAL("internal") }

enum class PersonEffect(val wire: String) {
    NONE("none"),
    RECEIVABLE_UP("receivable_up"), RECEIVABLE_DOWN("receivable_down"),
    PAYABLE_LOAN_UP("payable_loan_up"), PAYABLE_LOAN_DOWN("payable_loan_down"),
    PAYABLE_CUSTODY_UP("payable_custody_up"), PAYABLE_CUSTODY_DOWN("payable_custody_down"),
    BENEFICIARY_INFO("beneficiary_info"),
}

data class EconomicKindRule(
    val kind: EconomicKind,
    val labelKey: TextKey,
    val liquidity: Liquidity,
    /** بيزود الدخل الاقتصادي للفترة؟ */
    val countsAsIncome: Boolean,
    /** بيزود المصروف الشخصي؟ (نصيب المستخدم بس بعد فصل اللي على غيره) */
    val countsAsPersonalExpense: Boolean,
    val personEffect: PersonEffect,
) {
    /** الاسم المعروض باللغة الحالية (OVERRIDES §40) — بيتقرا وقت العرض مش وقت التحميل. */
    val label: String get() = uiText(labelKey)
}

private fun rule(kind: EconomicKind, labelKey: TextKey, liquidity: Liquidity, income: Boolean, expense: Boolean, effect: PersonEffect) =
    kind to EconomicKindRule(kind, labelKey, liquidity, income, expense, effect)

private val RULES: Map<EconomicKind, EconomicKindRule> =
    mapOf(
        rule(SALARY, TextKey.KIND_SALARY, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(BONUS, TextKey.KIND_BONUS, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(COMMISSION, TextKey.KIND_COMMISSION, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(OVERTIME, TextKey.KIND_OVERTIME, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(FREELANCE, TextKey.KIND_FREELANCE, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(PERSONAL_SALE, TextKey.KIND_PERSONAL_SALE, Liquidity.IN, true, false, PersonEffect.NONE),
        rule(LOAN_RECEIVED, TextKey.KIND_LOAN_RECEIVED, Liquidity.IN, false, false, PersonEffect.PAYABLE_LOAN_UP),
        rule(DEBT_COLLECTED, TextKey.KIND_DEBT_COLLECTED, Liquidity.IN, false, false, PersonEffect.RECEIVABLE_DOWN),
        rule(CUSTODY_RECEIVED, TextKey.KIND_CUSTODY_RECEIVED, Liquidity.IN, false, false, PersonEffect.PAYABLE_CUSTODY_UP),
        rule(PURCHASE, TextKey.KIND_PURCHASE, Liquidity.OUT, false, true, PersonEffect.RECEIVABLE_UP),
        rule(SUPPORT_GIFT, TextKey.KIND_SUPPORT_GIFT, Liquidity.OUT, false, true, PersonEffect.BENEFICIARY_INFO),
        rule(FEE, TextKey.KIND_FEE, Liquidity.OUT, false, true, PersonEffect.NONE),
        rule(LOAN_GRANTED, TextKey.KIND_LOAN_GRANTED, Liquidity.OUT, false, false, PersonEffect.RECEIVABLE_UP),
        rule(DEBT_REPAID, TextKey.KIND_DEBT_REPAID, Liquidity.OUT, false, false, PersonEffect.PAYABLE_LOAN_DOWN),
        rule(CUSTODY_RETURNED, TextKey.KIND_CUSTODY_RETURNED, Liquidity.OUT, false, false, PersonEffect.PAYABLE_CUSTODY_DOWN),
        rule(INTERNAL_TRANSFER, TextKey.KIND_INTERNAL_TRANSFER, Liquidity.INTERNAL, false, false, PersonEffect.NONE),
        rule(ASSET_BUY, TextKey.KIND_ASSET_BUY, Liquidity.OUT, false, false, PersonEffect.NONE),
        // بيع الأصل: الربح المحقق بس هو المكسب، ويتحسب في الاستثمار — مش دخل معيشة
        rule(ASSET_SELL, TextKey.KIND_ASSET_SELL, Liquidity.IN, false, false, PersonEffect.NONE),
        rule(UNCLASSIFIED, TextKey.KIND_UNCLASSIFIED, Liquidity.OUT, false, false, PersonEffect.NONE),
    )

fun ruleFor(kind: EconomicKind): EconomicKindRule = RULES.getValue(kind)

val ALL_ECONOMIC_KINDS: List<EconomicKind> = EconomicKind.entries.toList()

fun countsAsIncome(kind: EconomicKind): Boolean = ruleFor(kind).countsAsIncome
fun countsAsPersonalExpense(kind: EconomicKind): Boolean = ruleFor(kind).countsAsPersonalExpense

/** للتحذير في المراجعة، مش للرفض — الكشف بيعرف الاتجاه بس مش النية. */
fun isConsistentWithObservedDirection(kind: EconomicKind, observed: Direction): Boolean = when (ruleFor(kind).liquidity) {
    Liquidity.INTERNAL -> true
    Liquidity.IN -> observed == Direction.IN
    Liquidity.OUT -> observed == Direction.OUT
}
